File: 767-ReorganizeString/767-ReorganizeString.java
/*

767. Reorganize String
Medium

Given a string S, check if the letters can be rearranged so that two characters that are adjacent to each other are not the same.

If possible, output any possible result.  If not possible, return the empty string.

Example 1:

Input: S = "aab"
Output: "aba"

Example 2:

Input: S = "aaab"
Output: ""

Note:

    S will consist of lowercase letters and have length in range [1, 500].

*/

class Solution {
    public String reorganizeString(String S) {
        if(S == null || S.length() == 0 ) return "";
        char[] chArray = new char[S.length()];
        Map<Character, Integer> map = new HashMap<>();
        Queue<Pair> heap = new PriorityQueue<>((n1, n2) -> n2.getCount() - n1.getCount()); //max heap
        for(char ch : S.toCharArray())
            map.put(ch, map.containsKey(ch) ? map.get(ch) + 1 : 1);
        for(Map.Entry<Character, Integer> entry : map.entrySet())
            heap.add(new Pair(entry.getKey(), entry.getValue()));
        if(heap.peek().getCount() >= Math.ceil((double)S.length() / (double)2) + 1) return "";
        int idx = 0;
        while(!heap.isEmpty()){
            Pair p = heap.poll();
            while(p.getCount() != 0){
                chArray[idx] = p.getCh();
                idx = idx + 2;
                p.setCount(p.getCount() - 1);
                if(idx >= chArray.length)
                    idx = 1;
            }
        }
        return new String(chArray);
    }
}
class Pair{
    char ch;
    int count;
    Pair(char ch, int count){
        this.ch = ch;
        this.count = count;
    }
    public char getCh(){
        return ch;
    }
    public int getCount(){
        return count;
    }
    public void setCount(int count){
        this.count = count;
    }
}

File: 767-ReorganizeString/767-ReorganizeString.py
"""

767. Reorganize String

Given a string s, rearrange the characters of s so that any two adjacent characters are not the same.

Return any possible rearrangement of s or return "" if not possible.

Example 1:

Input: s = "aab"
Output: "aba"

Example 2:

Input: s = "aaab"
Output: ""

Constraints:

    1 <= s.length <= 500
    s consists of lowercase English letters.

"""

from heapq import heappush, heappop
from collections import Counter
class Solution:
    def reorganizeString(self, s: str) -> str:
        heap = []
        count = Counter(s)
        res = ['' for _ in range(len(s))]
        idx = 0
        for k, v in count.items(): heappush(heap, [-v, k])
        top = heappop(heap)
        if -top[0] > (len(s) + 1) // 2: return ""
        heappush(heap, top)
        while heap:
            top = heappop(heap)
            while top[0] != 0:
                res[idx] = top[1]
                top[0] += 1
                idx += 2
                if idx >= len(s): idx = 1
        return ''.join(res)
